File: js/Shop.ts
class Shop{
    private _game:Game;
    private _availableShapes:Shape[];
    private _possibleShapes:Shape[];
    private _leftOffset:number;

    constructor(game:Game){
        this._game = game;
        this._possibleShapes = new Array();
        this.initShapes();

        this._availableShapes = new Array();
        this._leftOffset = this._game.width/ (Settings.NUMAVAILABLESHAPES + 2); // |***|
    }
    render(){
        for(let as of this._availableShapes){
            as.render();
        }
    }

    update(){
        if (this._availableShapes.length == 0) {
            this.fillShop();
        }

        for(let as of this._availableShapes){
            as.update();
            //dont render if over board
            as.isOverShop = this._checkIfOverShop(as);
        }
    }

    get availableShapes():Shape[]{
        return this._availableShapes;
    }

    set availableShapes(availableShapes:Shape[]){
        throw new TypeError('Implemnet me');
    }

    protected initShapes(){
        this._possibleShapes[0] = new Shape(this._game, Shape.EMPTY, undefined, new Color(0,0,0));
        this._possibleShapes[1] = new Shape(this._game, Shape.L, undefined, new Color(0,0,0));
        this._possibleShapes[2] = new Shape(this._game, Shape.TRI, undefined, new Color(0,0,0));
        this._possibleShapes[3] = new Shape(this._game, Shape.QUAD, undefined, new Color(0,0,0));
        this._possibleShapes[4] = new Shape(this._game, Shape.BAR3HORIZONTAL, undefined, new Color(0,0,0));
        this._possibleShapes[5] = new Shape(this._game, Shape.BAR3VERTICAL, undefined, new Color(0,0,0));
        this._possibleShapes[6] = new Shape(this._game, Shape.J, undefined, new Color(0,0,0));
        this._possibleShapes[7] = new Shape(this._game, Shape.FLASH, undefined, new Color(0,0,0));
        this._possibleShapes[8] = new Shape(this._game, Shape.Z, undefined, new Color(0,0,0));
        this._possibleShapes[9] = new Shape(this._game, Shape.G , undefined, new Color(0,0,0));
        this._possibleShapes[10] = new Shape(this._game, Shape.DOT , undefined, new Color(0,0,0));
    }

    protected fillShop(){

        console.log(`fill up shop`);
        this._availableShapes = new Array();
        for(let i = 0; i < Settings.NUMAVAILABLESHAPES; i++ ){
            let random = Math.floor((Math.random() * (this._possibleShapes.length - 1) + 1 ));
            //console.log(this._leftOffset);
            this._availableShapes[i] = Shape.getClone(
                this._possibleShapes[random],
                 {x: (this._game.width/Settings.NUMAVAILABLESHAPES * i) + this._leftOffset,
                y:(this._game.height - (this._game.height - this._game.width)/2)}
                );
        }
        this._game.checkGameOver();
    }

    protected _checkIfOverShop(shape:Shape):boolean{
        return (shape.position.y > this._game.width);
    }

    touchStart(fingerPos:{x:number, y:number}):void{
        for (let s of this._availableShapes) {
            if(s.isTouched(fingerPos)){
                this._game.currentShape = s;
                //console.warn(`\n${this._game.currentShape.toString()}`);
                return;
            }
        }
    }

    deleteShape(currentShape:Shape):void{
        let index = this._availableShapes.indexOf(currentShape,0);
        console.log(`deleted ${index}`);
        if (index > -1) {
            this._availableShapes.splice(index, 1);
        }
        
        //this._availableShapes.splice(index, 1);
        //delete this._availableShapes[indexOfCurrentShape];
        console.log(`_availableShapes.length ${this._availableShapes.length}`);
    }
}

File: web/src/model/shop.dart
import 'dart:math';

import 'color.dart';
import 'game.dart';
import 'shape.dart';

class Shop {
  num NUM_AVAILABLE_SHAPES;
  List<Shape> availableShapes, possibleShapes;
  num leftOffset;

  Game game;

  Shop({this.game}) {
    possibleShapes = [];
    _iniShapes();
    availableShapes = [];
    leftOffset = game.width / NUM_AVAILABLE_SHAPES + 2;
  }

  void touchStart(Point<num> t) {}

  void update() {
    if (availableShapes.isEmpty) {
      _fillShop();
    }
    availableShapes.forEach((s) {
      s.update();
      //dont render if over board
      s.isOverShop = _checkIfOverShop(s);
    });
  }

  void render() {
    availableShapes.forEach((s) => s.render());
  }

  void _iniShapes() {
    // var possibleShapes = [];
    possibleShapes.add(Shape(
      game: game,
      pattern: Shape.EMPTY,
      position: Point(0, 0),
      color: Color(r: 0, g: 0, b: 0, a: 0),
    ));
    possibleShapes.add(Shape(
      game: game,
      pattern: Shape.L,
      position: Point(0, 0),
      color: Color(r: 0, g: 0, b: 0, a: 0),
    ));
    possibleShapes.add(Shape(
      game: game,
      pattern: Shape.TRI,
      position: Point(0, 0),
      color: Color(r: 0, g: 0, b: 0, a: 0),
    ));
    possibleShapes.add(Shape(
      game: game,
      pattern: Shape.QUAD,
      position: Point(0, 0),
      color: Color(r: 0, g: 0, b: 0, a: 0),
    ));
    possibleShapes.add(Shape(
      game: game,
      pattern: Shape.BAR3HORIZONTAL,
      position: Point(0, 0),
      color: Color(r: 0, g: 0, b: 0, a: 0),
    ));
    possibleShapes.add(Shape(
      game: game,
      pattern: Shape.BAR3VERTICAL,
      position: Point(0, 0),
      color: Color(r: 0, g: 0, b: 0, a: 0),
    ));
    possibleShapes.add(Shape(
      game: game,
      pattern: Shape.J,
      position: Point(0, 0),
      color: Color(r: 0, g: 0, b: 0, a: 0),
    ));
    possibleShapes.add(Shape(
      game: game,
      pattern: Shape.FLASH,
      position: Point(0, 0),
      color: Color(r: 0, g: 0, b: 0, a: 0),
    ));
    possibleShapes.add(Shape(
      game: game,
      pattern: Shape.Z,
      position: Point(0, 0),
      color: Color(r: 0, g: 0, b: 0, a: 0),
    ));
    possibleShapes.add(Shape(
      game: game,
      pattern: Shape.G,
      position: Point(0, 0),
      color: Color(r: 0, g: 0, b: 0, a: 0),
    ));
    possibleShapes.add(Shape(
      game: game,
      pattern: Shape.DOT,
      position: Point(0, 0),
      color: Color(r: 0, g: 0, b: 0, a: 0),
    ));
  }

  void _fillShop() {
    availableShapes = <Shape>[];
    for (var i = 0; i < NUM_AVAILABLE_SHAPES; i++) {
      var random = Random().nextInt(possibleShapes.length);
      availableShapes[i] = Shape.getClone(
          possibleShapes[random],
          Point(
            (game.width / NUM_AVAILABLE_SHAPES * i) + leftOffset,
            (game.height - (game.height - game.width) / 2),
          ));
    }
    game.checkGameOver();
  }

  bool _checkIfOverShop(Shape shape) {
    return (shape.position.y > game.width);
  }
}
